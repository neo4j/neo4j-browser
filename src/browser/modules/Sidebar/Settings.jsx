/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import { connect } from 'react-redux'
import * as actions from 'shared/modules/settings/settingsDuck'
import {
  Drawer,
  DrawerBody,
  DrawerHeader,
  DrawerSection,
  DrawerSectionBody,
  DrawerSubHeader
} from 'browser-components/drawer'
import { RadioSelector, CheckboxSelector } from 'browser-components/Form'
import {
  StyledSetting,
  StyledSettingLabel,
  StyledSettingTextInput
} from './styled'
import { toKeyString } from 'services/utils'
import {
  getExperimentalFeatures,
  experimentalFeatureSelfName,
  enableExperimentalFeature,
  disableExperimentalFeature
} from 'shared/modules/experimentalFeatures/experimentalFeaturesDuck'
import FeatureToggle from 'browser/modules/FeatureToggle/FeatureToggle'

const visualSettings = [
  {
    title: 'User Interface',
    settings: [
      {
        theme: {
          tooltip:
            'Use "Auto" to have neo4j-browser detect system dark vs. light mode if available.',
          displayName: 'Theme',
          type: 'radio',
          options: [
            actions.AUTO_THEME,
            actions.LIGHT_THEME,
            actions.OUTLINE_THEME,
            actions.DARK_THEME
          ]
        }
      },
      {
        codeFontLigatures: {
          displayName: 'Code font ligatures',
          tooltip: 'Use font ligatures for the command bar and cypher snippets',
          type: 'checkbox'
        }
      },
      {
        editorAutocomplete: {
          displayName: 'Enhanced query editor',
          tooltip: 'Autocomplete and syntax highlighting in query editor',
          type: 'checkbox'
        }
      },
      {
        enableMultiStatementMode: {
          displayName: 'Enable multi statement query editor',
          tooltip: 'Allows query editor to execute multiple statements',
          type: 'checkbox'
        }
      }
    ]
  },
  {
    title: 'Preferences',
    settings: [
      {
        showSampleScripts: {
          displayName: 'Show sample scripts',
          tooltip: 'Show sample scripts in favorites drawer.',
          type: 'checkbox'
        }
      },
      {
        initCmd: {
          displayName: 'Initial command to execute',
          tooltip: 'This commands is executed once connected to a graph.',
          type: 'input'
        }
      },
      {
        connectionTimeout: {
          displayName: 'Connection timeout (ms)',
          tooltip:
            'The timeout in milliseconds when establishing a connection to Neo4j.',
          type: 'input'
        }
      }
    ]
  },
  {
    title: 'Result Frames',
    settings: [
      {
        maxFrames: {
          displayName: 'Maximum number of result frames',
          tooltip:
            'Max number of result frames. When reached, old frames gets retired.'
        }
      },
      {
        maxHistory: {
          displayName: 'Max History',
          tooltip:
            'Max number of history entries. When reached, old entries gets retired.'
        }
      },
      {
        scrollToTop: {
          displayName: 'Scroll To Top',
          tooltip: 'Automatically scroll stream to top on new frames.',
          type: 'checkbox'
        }
      }
    ]
  },
  {
    title: 'Graph Visualization',
    settings: [
      {
        initialNodeDisplay: {
          displayName: 'Initial Node Display',
          tooltip:
            'Limit number of nodes displayed on first load of the graph visualization.'
        }
      },
      {
        maxNeighbours: {
          displayName: 'Max Neighbours',
          tooltip: 'Limit exploratory queries to this limit.'
        }
      },
      {
        maxRows: {
          displayName: 'Max Rows',
          tooltip: "Max number of rows to render in 'Rows' result view"
        }
      },
      {
        autoComplete: {
          displayName: 'Connect result nodes',
          tooltip:
            'If this is checked, after a cypher query result is retrieved, a second query is executed to fetch relationships between result nodes.',
          type: 'checkbox'
        }
      }
    ]
  }
]

export const Settings = ({
  settings,
  visualSettings,
  experimentalFeatures = {},
  onSettingsSave = () => {},
  onFeatureChange
}) => {
  if (!settings) return null
  const mappedSettings = visualSettings.map(visualSetting => {
    const title = <DrawerSubHeader>{visualSetting.title}</DrawerSubHeader>
    const mapSettings = visualSetting.settings
      .map(settingObj => {
        const setting = Object.keys(settingObj)[0]
        if (typeof settings[setting] === 'undefined') return null
        const visual = settingObj[setting].displayName
        const tooltip = settingObj[setting].tooltip || ''
        const type = settingObj[setting].type || 'input'

        if (type === 'input') {
          return (
            <StyledSetting key={toKeyString(visual)}>
              <StyledSettingLabel title={tooltip}>
                {visual}
                <StyledSettingTextInput
                  onChange={event => {
                    settings[setting] = event.target.value
                    onSettingsSave(settings)
                  }}
                  defaultValue={settings[setting]}
                  title={[tooltip]}
                  className={setting}
                />
              </StyledSettingLabel>
            </StyledSetting>
          )
        }

        if (type === 'radio') {
          return (
            <StyledSetting key={toKeyString(visual)}>
              <StyledSettingLabel title={tooltip}>{visual}</StyledSettingLabel>
              <RadioSelector
                options={settingObj[setting].options}
                onChange={event => {
                  settings[setting] = event.target.value
                  onSettingsSave(settings)
                }}
                selectedValue={settings[setting]}
              />
            </StyledSetting>
          )
        }

        if (type === 'checkbox') {
          return (
            <StyledSetting key={toKeyString(visual)}>
              <StyledSettingLabel title={tooltip}>
                <CheckboxSelector
                  onChange={event => {
                    settings[setting] = event.target.checked
                    onSettingsSave(settings)
                  }}
                  checked={settings[setting]}
                  data-testid={setting}
                />
                {visual}
              </StyledSettingLabel>
            </StyledSetting>
          )
        }
      })
      .filter(setting => setting !== null)

    return (
      <React.Fragment key={toKeyString(visualSetting.title)}>
        {title}
        {mapSettings}
      </React.Fragment>
    )
  })

  const mappedExperimentalFeatures = Object.keys(experimentalFeatures)
    .map(key => {
      const feature = experimentalFeatures[key]
      // Don't show the toggle to disable this section
      if (feature.name === experimentalFeatureSelfName) {
        return null
      }
      const visual = feature.displayName
      const tooltip = feature.tooltip || ''
      return (
        <StyledSetting key={toKeyString(feature.name)}>
          <CheckboxSelector
            onChange={event => {
              const on = event.target.checked
              onFeatureChange(feature.name, on)
            }}
            checked={experimentalFeatures[feature.name].on}
          />
          <StyledSettingLabel title={tooltip}>{visual}</StyledSettingLabel>
        </StyledSetting>
      )
    })
    .filter(r => r)

  return (
    <Drawer id="db-settings">
      <DrawerHeader>Browser Settings</DrawerHeader>
      <DrawerBody>
        <DrawerSection>
          <DrawerSectionBody key="settings">{mappedSettings}</DrawerSectionBody>
          <FeatureToggle
            name={experimentalFeatureSelfName}
            on={
              <>
                {mappedExperimentalFeatures.length ? (
                  <DrawerSubHeader>Experimental features</DrawerSubHeader>
                ) : null}
                <DrawerSectionBody key="experimental-features">
                  {mappedExperimentalFeatures}
                </DrawerSectionBody>
              </>
            }
          />
        </DrawerSection>
      </DrawerBody>
    </Drawer>
  )
}

const mapStateToProps = state => {
  return {
    experimentalFeatures: getExperimentalFeatures(state),
    settings: state.settings,
    visualSettings
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onSettingsSave: settings => {
      dispatch(actions.update(settings))
    },
    onFeatureChange: (name, on) => {
      if (on) {
        dispatch(enableExperimentalFeature(name))
      } else {
        dispatch(disableExperimentalFeature(name))
      }
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Settings)
