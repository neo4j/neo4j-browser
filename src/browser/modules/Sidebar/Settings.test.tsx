/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { render } from '@testing-library/react'
import React from 'react'

import { Settings } from './Settings'

test('Settings renders with strange characters in display name', () => {
  // Given
  const settings = { testSetting: true }
  const visualSettings = [
    {
      title: 'Test åäö settings',
      settings: [
        {
          testSetting: {
            displayName: 'åäö üüü'
          }
        }
      ]
    }
  ]

  // When
  const { container } = render(
    <Settings
      settings={settings}
      visualSettings={visualSettings}
      telemetrySettings={{
        allowUserStats: false,
        allowCrashReporting: false,
        source: 'BROWSER_SETTING'
      }}
    />
  )

  // Then
  expect(container).toMatchSnapshot()
})
