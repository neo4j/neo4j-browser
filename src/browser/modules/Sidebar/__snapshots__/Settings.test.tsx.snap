// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings renders with strange characters in display name 1`] = `
<div>
  <div
    class="sc-ckVGcZ dbUHn"
    id="db-settings"
  >
    <h4
      class="sc-jKJlTe hMgJZm"
    >
      Browser Settings
    </h4>
    <div
      class="sc-iAyFgw mmJws"
    >
      <div
        class="sc-kEYyzF fIIWFC"
      >
        <div
          class="sc-kkGfuU fqDSiG"
        >
          <h5
            class="sc-hMqMXs eRgQj"
          >
            Test åäö settings
          </h5>
          <div
            class="sc-jWBwVP cLuGZv"
          >
            <label
              class="sc-brqgnP jpgGAL"
              title=""
            >
              åäö üüü
              <input
                class="sc-jAaTju ePWwEZ testSetting"
                data-testid="setting-testSetting"
                title=""
                value="true"
              />
            </label>
          </div>
          <h5
            class="sc-hMqMXs eRgQj"
          >
            Product Analytics
          </h5>
        </div>
        <div
          class="sc-kkGfuU fqDSiG"
        />
      </div>
    </div>
  </div>
</div>
`;
